DOCS-1524: Add PID config to motors and bases (#2547)
viamrobotics/docs: docs/components/base/sensor-controlled.md

A `sensor-controlled` base supports a robotic base with feedback control from a movement sensor.

{{% alert title="Requirements" color="note" %}}

> In order to use feedback control for [SetVelocity()](/components/base/#setvelocity), you must provide a movement sensor that implements [Spin()](/components/base/#getangularvelocity) and [Spin()](/components/base/#getlinearvelocity).

Reviewer: ```suggestion
In order to use feedback control for [SetVelocity()](/components/base/#setvelocity), you must provide a movement sensor that implements [AngularVelocity()](/components/base/#getangularvelocity) and [LinearVelocity()](/components/base/#getlinearvelocity).
```

> In order to use feedback control for [SetVelocity()](/components/base/#setvelocity), you must provide a movement sensor that implements [AngularVelocity()](/components/base/#getangularvelocity) and [LinearVelocity()](/components/base/#getlinearvelocity).

Author: ```suggestion
In order to use feedback control for [SetVelocity()](/components/movement-sensor/#setvelocity), you must provide a movement sensor that implements [AngularVelocity()](/components/movement-sensor/#getangularvelocity) and [LinearVelocity()](/components/movement-sensor/#getlinearvelocity).
```

> In order to use feedback control for [SetVelocity()](/components/movement-sensor/#setvelocity), you must provide a movement sensor that implements [AngularVelocity()](/components/movement-sensor/#getangularvelocity) and [LinearVelocity()](/components/movement-sensor/#getlinearvelocity).

Author: ```suggestion
In order to use feedback control for [SetVelocity()](/components/base/#setvelocity), you must provide a movement sensor that implements [AngularVelocity()](/components/movement-sensor/#getangularvelocity) and [LinearVelocity()](/components/movement-sensor/#getlinearvelocity).
```

> In order to use feedback control for [Spin()](/components/base/#spin), you must provide a movement sensor that implements [Spin()](/components/base/#getorientation).

Reviewer: ```suggestion
In order to use feedback control for [Spin()](/components/base/#spin), you must provide a movement sensor that implements [Orientation()](/components/base/#getorientation).
```

> In order to use feedback control for [Spin()](/components/base/#spin), you must provide a movement sensor that implements [Orientation()](/components/base/#getorientation).

Author: ```suggestion
In order to use feedback control for [Spin()](/components/base/#spin), you must provide a movement sensor that implements [Orientation()](/components/movement-sensor/#getorientation).
```

<!-- prettier-ignore -->
| Name | Type | Inclusion | Description |
| ---- | ---- | --------- | ----------- |
| `movement_sensor` | array | **Required** | Array with the `name`s of any movement sensors on your base you want to gather feedback from. The driver will select the first movement sensor providing appropriate feedback for either the `SetVelocity()` or the `Spin()` endpoint. |
| `base` | string | **Required** | String with the `name` of the base you want to wrap with sensor control. |

> | `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integral, and derivative terms for linear and angular velocity. If you want these values to be auto-tuned, you can set all values to 0: `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, and `viam-server` will auto-tune and log the calculated values. Tuning takes several seconds and spins the motors. Copy the values from the logs and add them to the configuration once tuned for the values to take effect. For more information see [Feedback contrel](#feedback-control). |

Author: ```suggestion
| `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integral, and derivative terms for linear and angular velocity. If you want these values to be auto-tuned, you can set all values to 0: `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, and `viam-server` will auto-tune and log the calculated values. Tuning takes several seconds and spins the motors. Copy the values from the logs and add them to the configuration once tuned for the values to take effect. If you need to auto-tune multiple controlled components that depend on the same hardware, such as a sensor controlled base and one of the motors on the base, run the auto-tuning process one component at a time. For more information see [Feedback contrel](#feedback-control). |
```

| ---- | ---- | --------- | ----------- |

> | `movement-sensor` | array | **Required** | Array with the `name`s of any movement sensors on your base you want to gather feedback from. The driver will select the first movement sensor providing appropriate feedback for either the `SetVelocity()` or the `Spin()` endpoint. |

Reviewer: Also noticed this typo: I think it should be `movement_sensor` (underscore not dash)

> | `movement-sensor` | array | **Required** | Array with the `name`s of any movement sensors on your base you want to gather feedback from. The driver will select the first movement sensor providing appropriate feedback for either the `SetVelocity()` or the `Spin()` endpoint. |

Author: ```suggestion
| `movement_sensor` | array | **Required** | Array with the `name`s of any movement sensors on your base you want to gather feedback from. The driver will select the first movement sensor providing appropriate feedback for either the `SetVelocity()` or the `Spin()` endpoint. |
```

| `base` | string | **Required** | String with the `name` of the base you want to wrap with sensor control. |

> | `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integrative, and derivative terms for linear and angular velocity. If you use `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, `viam-server` auto-tunes the values and logs the values. To avoid retuning the control parameters constantly, copy the values from the log and add them to the configuration once tuned. For more information see [Control base velocity with a sensor](#control-base-velocity-with-a-sensor). |

Author: ```suggestion
| `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integrative, and derivative terms for linear and angular velocity. If you use `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, `viam-server` auto-tunes the values and logs the values. Tuning takes several seconds and spins the motors. To avoid tuning every time the robot starts up, copy the values from the log and add them to the configuration once tuned. For more information see [Control base velocity with a sensor](#control-base-velocity-with-a-sensor). |
```

> | `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integrative, and derivative terms for linear and angular velocity. If you use `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, `viam-server` auto-tunes the values and logs the values. Tuning takes several seconds and spins the motors. To avoid tuning every time the robot starts up, copy the values from the log and add them to the configuration once tuned. For more information see [Control base velocity with a sensor](#control-base-velocity-with-a-sensor). |

Reviewer 2: integrative -> integral

Reviewer 2: maybe rephrase to say "if you want these values to be auto-tuned, you can set all values to 0 `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, and `viam-server` will  auto-tune, and log the calculated values"

> | `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integrative, and derivative terms for linear and angular velocity. If you use `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, `viam-server` auto-tunes the values and logs the values. Tuning takes several seconds and spins the motors. To avoid tuning every time the robot starts up, copy the values from the log and add them to the configuration once tuned. For more information see [Control base velocity with a sensor](#control-base-velocity-with-a-sensor). |

Reviewer 2: also maybe "comfy the values form the logs"

> | `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integral, and derivative terms for linear and angular velocity. If you want these values to be auto-tuned, you can set all values to 0: `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, and `viam-server` will auto-tune and logs the calculated values. Tuning takes several seconds and spins the motors. To avoid tuning every time the robot starts up, copy the values from the log and add them to the configuration once tuned. For more information see [Control base velocity with a sensor](#control-base-velocity-with-a-sensor). |

Reviewer 2: ```suggestion
| `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integral, and derivative terms for linear and angular velocity. If you want these values to be auto-tuned, you can set all values to 0: `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, and `viam-server` will auto-tune and log the calculated values. Tuning takes several seconds and spins the motors. To avoid tuning every time the robot starts up, copy the values from the logs and add them to the configuration once tuned. For more information see [Control base velocity with a sensor](#control-base-velocity-with-a-sensor). |
```

> | `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integral, and derivative terms for linear and angular velocity. If you want these values to be auto-tuned, you can set all values to 0: `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, and `viam-server` will auto-tune and log the calculated values. Tuning takes several seconds and spins the motors. To avoid tuning every time the robot starts up, copy the values from the logs and add them to the configuration once tuned. For more information see [Control base velocity with a sensor](#control-base-velocity-with-a-sensor). |

Reviewer: Do we want to add `control_parameters` to the config and JSON template examples too? I thought it was nice how we added it for the gpio motor

Reviewer: resurfacing this comment

> | `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integral, and derivative terms for linear and angular velocity. If you want these values to be auto-tuned, you can set all values to 0: `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, and `viam-server` will auto-tune and log the calculated values. Tuning takes several seconds and spins the motors. To avoid tuning every time the robot starts up, copy the values from the logs and add them to the configuration once tuned. For more information see [Control base velocity with a sensor](#control-base-velocity-with-a-sensor). |

Reviewer 2: good catch yes this is true, they need to be copied into the config to ever be applied

> | `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integral, and derivative terms for linear and angular velocity. If you want these values to be auto-tuned, you can set all values to 0: `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, and `viam-server` will auto-tune and log the calculated values. Tuning takes several seconds and spins the motors. To avoid tuning every time the robot starts up, copy the values from the logs and add them to the configuration once tuned. For more information see [Control base velocity with a sensor](#control-base-velocity-with-a-sensor). |

Author: ```suggestion
| `control_parameters` | object | Optional | A JSON object containing the coefficients for the proportional, integral, and derivative terms for linear and angular velocity. If you want these values to be auto-tuned, you can set all values to 0: `[ { "type": "linear_velocity", "p": 0, "i": 0, "d": 0 }, { "type": "angular_velocity", "p": 0, "i": 0, "d": 0 } ]`, and `viam-server` will auto-tune and log the calculated values. Tuning takes several seconds and spins the motors. Copy the values from the logs and add them to the configuration once tuned for the values to take effect. For more information see [Control base velocity with a sensor](#control-base-velocity-with-a-sensor). |
```

- [SetVelocity()](/components/base/#setvelocity): available if base is configured to receive angular and linear velocity feedback.

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Reviewer: Relatedly, @martha-johnston , I realized I'm actually confused what happens if the user specifies multiple movement sensors in the config. Does their output get merged somehow? When would a user want to do that rather than use a merged movement sensor?

Reviewer 2: it will go through all the sensors and return the first sensor that supports the property it's looking for. so first go through all the sensors and find one that supports orientation, then go through them all again and find one that supports linear and angular velocity

Reviewer: got it, makes sense!

Reviewer: no need to change this for this PR, but is there a reason to support this functionality separate from merged movement sensor? if not i could file a ticket for the RDK backlog to make this a string field

Reviewer: gotcha. i'll go ahead and file a ticket so that we can have a discussion about it with rand

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Reviewer 2: so spin still will do this as it has been it basically just doesn't use a PID loop. similarly, set velocity will still use sensor feedback but if you don't configure the control parameters it won't use a PID loop. if that makes sense

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Reviewer: ahh, i see

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Reviewer: (will read the page again with this context and either re-comment or resolve)

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Reviewer: Okay, I would suggest the following:

* Adjust the "Requirements" alert box to say something like:

In order to use feedback control for SetVelocity, you must provide a movement sensor that implements GetAngularVelocity and GetLinearVelocity.

In order to use feedback control for Spin, you must provide a movement sensor that implements GetOrientation.

* Remove the following text since it's very similar to what's in the Requirements alert box:

```
The following base control API methods are available with feedback control on a sensor-controlled base:

[SetVelocity()](https://docs-test.viam.dev/2547/components/base/#setvelocity): available if base is configured to receive angular and linear velocity feedback.
[Spin()](https://docs-test.viam.dev/2547/components/base/#spin): available if base is configured to receive orientation feedback.
```

* Maybe rename `Control base velocity with a sensor` to `Feedback control for SetVelocity` and move it above `Configure a sensor-controlled base as follows:`.

* Maybe add a section `Feedback control for Spin` below `Feedback control for SetVelocity`, and have it say something like:

Spin implements a basic form of feedback control that increases or decreases the power being set on the motors in increments of 10%.

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Reviewer 2: here I'm also not sure spin is doing the 10% increment. I'll take a closer look at the code and get back to you

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Reviewer: thank you!

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Reviewer 2: okay spin starts the motors at the specified degrees/sec by calling `SetVelocity` at that speed on the underlying base. then it does some math and monitors how far the base has spun and stops once it has reached the goal position

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Reviewer 2: this is only if the desired degree < 360. any spin call for more than 360 deg just calls the underlying base's Spin

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Reviewer: got it, thanks for looking into it! updated the suggestion for `Feedback control for Spin`

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Author: Ok - made some changes, let me know what you think

> - [Spin()](/components/base/#spin): available if base is configured to receive orientation feedback.

Reviewer: Looks great!
